library/core/tests/iter/adapters/map_windows.rs

```diff
@@ -0,0 +1,313 @@
+        assert_eq!(iter.next(), Some([0, 1]));
+    }
+    assert_eq!(counter, 2);
+    {
+        let mut iter = (2..5).inspect(|_| counter += 1).map_windows(|arr: &[i32; 2]| *arr);
+        assert_eq!(iter.next(), Some([2, 3]));
+        assert_eq!(iter.next(), Some([3, 4]));
+    }
+    assert_eq!(counter, 5);
+}
+
+#[test]
+fn test_size_hint() {
+    struct SizeHintCheckHelper((usize, Option<usize>));
+
+    impl Iterator for SizeHintCheckHelper {
+        type Item = i32;
+
+        fn next(&mut self) -> Option<i32> {
+            Some(0)
+        }
+
+        fn size_hint(&self) -> (usize, Option<usize>) {
+            self.0
+        }
+    }
+
+    fn check_size_hint<const N: usize>(
+        input_size_hint: (usize, Option<usize>),
+        expected_size_hint: (usize, Option<usize>),
+    ) {
+        let iter = SizeHintCheckHelper(input_size_hint).map_windows(|_: &[_; N]| ());
+        assert_eq!(iter.size_hint(), expected_size_hint);
+    }
+
+    check_size_hint::<1>((0, None), (0, None));
+    check_size_hint::<1>((0, Some(0)), (0, Some(0)));
+    check_size_hint::<1>((0, Some(2)), (0, Some(2)));
+    check_size_hint::<1>((1, None), (1, None));
+    check_size_hint::<1>((1, Some(1)), (1, Some(1)));
+    check_size_hint::<1>((1, Some(4)), (1, Some(4)));
+    check_size_hint::<1>((5, None), (5, None));
+    check_size_hint::<1>((5, Some(5)), (5, Some(5)));
+    check_size_hint::<1>((5, Some(10)), (5, Some(10)));
+
+    check_size_hint::<2>((0, None), (0, None));
+    check_size_hint::<2>((0, Some(0)), (0, Some(0)));
+    check_size_hint::<2>((0, Some(2)), (0, Some(1)));
+    check_size_hint::<2>((1, None), (0, None));
+    check_size_hint::<2>((1, Some(1)), (0, Some(0)));
+    check_size_hint::<2>((1, Some(4)), (0, Some(3)));
+    check_size_hint::<2>((5, None), (4, None));
+    check_size_hint::<2>((5, Some(5)), (4, Some(4)));
+    check_size_hint::<2>((5, Some(10)), (4, Some(9)));
+
+    check_size_hint::<5>((0, None), (0, None));
+    check_size_hint::<5>((0, Some(0)), (0, Some(0)));
+    check_size_hint::<5>((0, Some(2)), (0, Some(0)));
+    check_size_hint::<5>((1, None), (0, None));
+    check_size_hint::<5>((1, Some(1)), (0, Some(0)));
+    check_size_hint::<5>((1, Some(4)), (0, Some(0)));
+    check_size_hint::<5>((5, None), (1, None));
+    check_size_hint::<5>((5, Some(5)), (1, Some(1)));
+    check_size_hint::<5>((5, Some(10)), (1, Some(6)));
+}
+
+#[test]
+fn test_advance_by() {
+    fn check_advance_by<I, const N: usize>(
+        iter: I,
+        n: usize,
+        advanced_result: Result<(), usize>,
+        before_advance: &[[I::Item; N]],
+        after_advance: &[[I::Item; N]],
+    ) where
+        I: Iterator,
+        I::Item: Copy + PartialEq + std::fmt::Debug,
+    {
+        // FIXME: SIGSEGV: invalid memory reference without `.filter_map(Some)`
+        let mut iter = iter.filter_map(Some).map_windows(|arr: &[_; N]| *arr);
+        // let mut iter = iter.map_windows(|arr: &[_; N]| *arr);
```

Comment:
Now I'm confused here why if I remove `.filter_map(Some)`, it crashes with `SIGSEGV: invalid memory reference`.

Reply:
It was caused by the wrong `.size_hint()`. When the buffer is initialized, the `.size_hint()` should be equal to the inner iterator's.